PR - Iris Disbursement API

Xaxxis opened the pull request:
New feature:
- API Iris disbursement
- Payment BIN API
- Set Connection Pool & Keep alive connection
- Set header X-Idempotency-key for Iris and Idempotency-Key for Payment
- Set `X-Append-Notification`: to add new notification url(s) alongside the settings on dashboard
- Set `X-Override-Notification`: to use new notification url(s) disregarding the settings on dashboard
- Library version on `User-Agent` header

Improvement:

- Change logic for setup headers
- Remove versioning path from base URL, move to `String API_VERSION` in CoreAPI, SnapAPI interface class
- Improve conditional logic for HttpLoggingInterceptors. Now all debugging response handle only from HttpLoggingInterceptors Which can show logs request and response information

Bugs fix:
- Failure/error response from Midtrans api is empty json.

Commit:
add maven

Commit:
Add changelog.md for tracking

Commit:
Update java docs

Commit:
Update demo app

- Update version lib for sample and heroku config at pom.xml
- Add sample Iris Disbursement
- Add Simulator Android SDK Mobile
- Add MobileSdkController sample

Commit:
Update library version to 1.1.2

Commit:
Update Config

- Added setter & getter max connection pool
- Added setter & getter keep alive durations
- Added setter & getter append/ovveride notif
- Added setter & getter Idempotent key Iris/Payment
- Added Iris API Base URL

Commit:
Update Config Builder

- Change default value , connection timeout, read, write
- Added default value max connection, keep alive durations

Commit:
Added Iris Api factory

Commit:
Update HttpClient

- Added connection pool configuration
- change Http headers setup, added some headers params
- change conditional logic LogingInterceptor
- added new method to get library version

Commit:
Added API Version on path api url

Commit:
error utils deprecated

Commit:
Revert "error utils deprecated"

This reverts commit cf8f42d6a3b25d688ab20fcb39cb32596f898154.

Commit:
ErorUtils deprecated

All API response failure/error will return to JSONObject

Commit:
New Client API for Iris Dishbursement

Commit:
Change path with api version

Commit:
change final int for port

Commit:
remove error utils, will handle from response errorBody

Commit:
Implment Iris API

Commit:
remove error utils, will handle from response errorBody

Commit:
Added BIN API

Commit:
Midtrans Iris API

Commit:
added version for user-agent

Commit:
added iris config scenario test

Commit:
CoreAPI failure scenario test

Commit:
Iris Testing api

Commit:
Update constant for testing purpose

Commit:
add iris data mockup for test

Commit:
Update Snap failure scenario test

Commit:
Update readme regarding the library update

Commit:
clean code

Commit:
Increasing major changes version Library and Sample

Commit:
Update docs iris set & get merchant-key

Commit:
added Iris set and get merchant-key for iris notification

Commit:
Added sample iris handling notification

Commit:
Merge branch 'master' into master

Commit:
Update version new library

Commit:
Remove unused comment

Commit:
Merge branch 'master' of https://github.com/Xaxxis/midtrans-java

rizdaprasetya commented on library/pom.xml:
```diff
@@ -80,13 +80,22 @@
 			<id>bintray-midtrans-midtrans-java</id>
 			<name>midtrans-midtrans-java</name>
-			<url>https://api.bintray.com/maven/midtrans/midtrans-java/com.midtrans/;publish=1</url>
+            <url>https://api.bintray.com/maven/midtrans/midtrans-java/com.midtrans/;publish=1</url>
```
indentation looks not consistent

Xaxxis replied:
done

rizdaprasetya commented on library/src/main/java/com/midtrans/Config.java:
```diff
@@ -8,19 +8,25 @@
     private String CLIENT_KEY;
-
     private boolean isProduction;
+    private String irisMerchantKey;
```
Why is it different from `SERVER_KEY` to `irisMerchantKey`? not  uppercase?

rizdaprasetya commented on library/src/main/java/com/midtrans/service/MidtransIrisApi.java:
```diff
@@ -0,0 +1,151 @@
+     * @return {JSONObject} - org.json Promise contains Object from JSON decoded response (more params detail refer to: https://iris-docs.midtrans.com/#list-banks)
+     * @throws MidtransError when an exception was occurred during executing the request.
+     */
+    JSONObject getListBank() throws MidtransError;
```
change to more descriptive `getBeneficiaryBanks`

rizdaprasetya commented on library/src/main/java/com/midtrans/service/MidtransIrisApi.java:
```diff
@@ -0,0 +1,151 @@
+     * @return {JSONObject} - org.json Promise contains Object from JSON decoded response
+     * @throws MidtransError when an exception was occurred during executing the request.
+     */
+    JSONObject getPayoutsDetails(String referenceNo) throws MidtransError;
```
`getPayoutDetails` payout without `s`

rizdaprasetya commented on library/src/main/java/com/midtrans/service/MidtransIrisApi.java:
```diff
@@ -0,0 +1,151 @@
+     * @return {JSONObject} - org.json Promise contains Object from JSON decoded response refer to: https://iris-docs.midtrans.com/#list-beneficiaries
+     * @throws MidtransError when an exception was occurred during executing the request.
+     */
+    JSONArray getListBeneficiaries() throws MidtransError;
```
change to `getBeneficiaries`

rizdaprasetya commented on library/src/main/java/com/midtrans/service/MidtransIrisApi.java:
```diff
@@ -0,0 +1,151 @@
+     * @return {JSONObject} - org.json Promise contains Object from JSON decoded response refer to: https://iris-docs.midtrans.com/#check-balance-aggregator
+     * @throws MidtransError when an exception was occurred during executing the request.
+     */
+    JSONObject getAggregatorBalance() throws MidtransError;
```
change to `getBalance`

rizdaprasetya reviewed (changes requested):
This is technical feedback, readme related feedback ill be submitted separately

rizdaprasetya commented on CHANGELOG.md:
```diff
@@ -0,0 +1,47 @@
+
+- Change logic for setup headers
+- Remove versioning path from base URL, move to `String API_VERSION` in CoreAPI, SnapAPI interface class
+- Improve conditional logic for HttpLoggingInterceptors. Now all debugging response handle only from HttpLoggingInterceptors Which can show logs request and response information
```
`will be handled only from`

rizdaprasetya commented on CHANGELOG.md:
```diff
@@ -0,0 +1,47 @@
+- Improve conditional logic for HttpLoggingInterceptors. Now all debugging response handle only from HttpLoggingInterceptors Which can show logs request and response information
+
+Bugs fix:
+- Failure/error response from Midtrans api is empty json. 
```
`Midtrans api result in empty json`

rizdaprasetya commented on CHANGELOG.md:
```diff
@@ -0,0 +1,47 @@
+
+Improvement:
+
+HttpLoggingInterceptor: Which can show logs request and response information. This feature auto disable when using production mode.
```
This feature is disabled by default on production mode.

rizdaprasetya commented on CHANGELOG.md:
```diff
@@ -0,0 +1,47 @@
+- Can use HTTP Proxy feature
+- Can set connectionTimeout
+- Can set readTimeout
+- Can set writeTimeout
```
- Allow HTTP Proxy config
- Allow set connectionTimeout
- Allow set readTimeout
- Allow set writeTimeout

rizdaprasetya commented on README.md:
````diff
@@ -87,7 +88,20 @@ MidtransSnapApi snapApi = new ConfigFactory(new Config("YOU_SERVER_KEY","YOUR_CL
+`isProduction`
+```
+
+You can also re-setting config using `apiConfig()` method on MidtransCoreApi.Class, MidtransSnapApi.Class or MidtransIrisApi.class like `coreApi.apiConfig.set( ... )`
````
can also re-set config

rizdaprasetya commented on README.md:
````diff
@@ -99,9 +113,15 @@ coreApi.apiConfig().setSERVER_KEY("YOUR_SERVER_KEY");
 ```
 
-In production environment, LOG is automatically turn off, if you want enable LOG, you can turn on log with this
+In production environment, LOG is automatically turn off, if you want enable LOG, you can `setEnabledLog` like this
````
`LOG is by default turned off`
`you can enable by setEnabledLog, e.g:`

rizdaprasetya commented on README.md:
````diff
@@ -514,12 +557,192 @@ private Map<String,String> params() {
     //do something with `result` JSON Raw Object
 ```
+### 2.2.D Iris API (Disbursement)
+You can see some Iris API examples [here](example/src/main/java/com/midtrans/sample/controller/IrisController.java).
````
remove `some`

rizdaprasetya commented on library/src/main/java/com/midtrans/service/MidtransIrisApi.java:
```diff
@@ -0,0 +1,151 @@
+public interface MidtransIrisApi {
+
+    /**
+     * Do re-setting config Class iris-credential, IrisIdempotencyKey
```
`re-set config Class`

rizdaprasetya reviewed (changes requested):
Feedback for readme related

Commit:
Update from feedback for readme related

Commit:
Reformat code with IDE to fix identation

Commit:
deleted unused class

Commit:
Update from technical feedback

- Change irisMerchantKey to Uppercase

Change Method name:
- `getAgregatorBalance` to` getBalance`
- `getListBeneficiaries` to `getBenefiaries`
- `getPayoutsDetails` to `getPayoutDetails`
- `getListBank` to getBeneficaryBanks`

Change wording re-setting to re-set

Commit:
Update java docs regarding technical feedback

Commit:
Change method name on HttpClient class regarding technical update

Commit:
Update java docs

rizdaprasetya reviewed (approved):
Looks good now! please merge and release! Thanks @Xaxxis
